File: Neetcode/DailyTemp.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<int> temp, index;
        vector<int> ans(temperatures.size());
        for (int i=0; i<temperatures.size(); i++){
            if (temp.empty()){
                temp.push(temperatures[i]);
                index.push(i);
            }
            else{
                if (temperatures[i] <= temp.top()){
                    temp.push(temperatures[i]);
                    index.push(i);
                }
                else{
                    while (!temp.empty() && temperatures[i] > temp.top()){
                        temp.pop();
                        ans[index.top()] = i - index.top();
                        index.pop();
                    }
                    temp.push(temperatures[i]);
                    index.push(i);
                }
            }
        }
        return ans;
    }
};

File: Neetcode/DailyTemp.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []
        index = []
        res = [0]*len(temperatures)
        for i,t in enumerate(temperatures):
            if stack:
                if t <= stack[-1]:
                    stack.append(t)
                    index.append(i)
                else:
                    while stack and (t > stack[-1]):
                        res[index[-1]] = i-index[-1]
                        stack.pop()
                        index.pop()
                    stack.append(t)
                    index.append(i)
            else:
                stack.append(t)
                index.append(i)
        return res
